src/descriptor/satisfied_contraints.rs

```diff
@@ -0,0 +1,1482 @@
+            &Descriptor::ShWpkh(ref pk) |
+            &Descriptor::Wpkh(ref pk) => SatisfiedConstraints{
+                secp,
+                sighash,
+                public_key: Some(&pk),
+                state : vec![],
+                stack,
+                altstack: vec![] as Vec<StackElement<'stack>>,
+                age,
+                height,
+            },
+            &Descriptor::Sh(ref miniscript) |
+            &Descriptor::Bare(ref miniscript) |
+            &Descriptor::ShWsh(ref miniscript) |
+            &Descriptor::Wsh(ref miniscript) => SatisfiedConstraints{
+                secp,
+                sighash,
+                public_key: None,
+                state : vec![(&miniscript.to_inner() , NodeEvaluationState::NoChildrenCompleted)],
+                stack,
+                altstack: vec![] as Vec<StackElement<'stack>>,
+                age,
+                height,
+            }
+        }
+    }
+}
+
+impl<'desc, 'stack, Pk, Pkh> Iterator for SatisfiedConstraints<'desc, 'stack, Pk, Pkh>
+where Pk: Clone + ToPublicKey, Pkh: ToPublicKeyHash + Clone,
+{
+    type Item = Result<SatisfiedConstraint<'desc, 'stack, Pk, Pkh>, Error>;
+
+    fn next(&mut self) -> Option< Result<SatisfiedConstraint<'desc, 'stack, Pk, Pkh>, Error>> {
+        //state empty implies that either the execution has terminated or we have a
+        //Pk based descriptor
+        if self.state.is_empty() {
+            if let Some(pk) = self.public_key{
+                if let Some(StackElement::Push(sig)) = self.stack.pop(){
+                    if let Ok(sig) = verify_sersig(self.secp, self.sighash, &pk.to_public_key(), sig){
+                        //Signature check successful, set public_key to None to
+                        //terminate the next() function in the subsequent call
+                            self.public_key = None;
+                            self.stack.push(StackElement::Satisfied);
+                           return Some(Ok(SatisfiedConstraint::PublicKey { key: pk, sig }))
+                        } else{
+                            return Some(Err
+                                (Error::CouldnotEvaluate("Signature failure".to_owned())))
+                        }
+                    } else{
+                        return Some(Err(CouldnotEvaluate("Unexpected Stack End".to_owned())))
+                    }
+                } else {
+                //All the script has been executed.
+                //Check that the stack must contain exactly 1 satisfied element
+                if self.stack.pop() == Some(StackElement::Satisfied)
+                    && self.stack.is_empty() {
+                    return None
+                } else {
+                    return Some(Err(
+                        Error::CouldnotEvaluate("Script Is not satisfied".to_owned())))
+                }
+            }
+        }
+
+        let (root, node_state) = self.state.pop().unwrap();//non-empty stack
```

Comment:
I think it would be much simpler to do

```
while let Some((root, node_state)) = self.state.pop() {
    // ...
}
```
rather than tail-call recursing here. Then you could move the above "empty stack" check to after the loop, replace all your `Some(x)` with `return Some(x)`, and all your `None` with nothing.